edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,326 @@
+                this.nodes
+                .Select(kvp => kvp.Value)
+                .Where(serviceIdentity => serviceIdentity.Identity.IsEdgeDevice && serviceIdentity.Identity.DeviceScope.Contains(scopeId))
+                .ToList();
+
+            if (devices.Count() > 0)
+            {
+                return Option.Some(devices.First());
+            }
+
+            return Option.None<ServiceIdentityTreeNode>();
+        }
+
+        /// <summary>
+        /// This is an object to wrap ServiceIdentity so that it can
+        /// point to a parent and act as a node in a tree.
+        /// </summary>
+        internal class ServiceIdentityTreeNode
+        {
+            // Only allow up to 5 Edge devices to be linked together
+            static readonly int MaxNestingDepth = 4;
+
+            public ServiceIdentity Identity { get; }
+            public Option<string> AuthChain { get; private set; }
+            public Option<ServiceIdentityTreeNode> Parent { get; private set; }
+
+            List<ServiceIdentityTreeNode> children;
+            int currentDepth;
+
+            public ServiceIdentityTreeNode(ServiceIdentity identity, Option<string> authChain)
+            {
+                this.Identity = Preconditions.CheckNotNull(identity);
+                this.Parent = Option.None<ServiceIdentityTreeNode>();
+                this.children = new List<ServiceIdentityTreeNode>();
+                this.AuthChain = authChain;
+                this.currentDepth = 0;
+            }
+
+            public void AddChild(ServiceIdentityTreeNode childNode)
+            {
+                if (!this.Identity.IsEdgeDevice)
+                {
+                    throw new ArgumentException($"{this.Identity.Id} is not an Edge device, only Edge devices can have children");
+                }
+
+                if (this.currentDepth >= MaxNestingDepth && childNode.Identity.IsEdgeDevice)
+                {
+                    // Don't add Edge devices greater than the max depth.
+                    // If the customer configured more layers than the max,
+                    // we'll just ignore anything beyond the max and
+                    // continue working with the what we have.
+                    Events.MaxDepthExceeded(childNode.Identity.Id);
+                    return;
+                }
+
+                this.children.Add(childNode);
+                childNode.Parent = Option.Some(this);
+                childNode.UpdateAuthChainFromParent(this, this.currentDepth);
+            }
+
+            public void RemoveChild(ServiceIdentityTreeNode childNode)
+            {
+                this.children.Remove(childNode);
+                childNode.Parent = Option.None<ServiceIdentityTreeNode>();
+                childNode.RemoveAuthChain();
+            }
+
+            public void RemoveAllChildren()
+            {
+                var snapshot = new List<ServiceIdentityTreeNode>(this.children);
```

Comment:
can it just loop through the list and update the child and then set this.children to empty list?

Reply:
I don't think that saves us anything, right?  It'd still need to go through and recursively remove the authchain on each grandchildren, etc.